### PR TITLE
Add table iterators

### PR DESCRIPTION
**Pull request by drwells**
Follow-up to #5959: this commit adds iterator interfaces to `Table` (and, subsequently, `FullMatrix`) based on the work done in the first PR.

I also went ahead and cleaned up our `table.h` inclusions: there were a lot of places where we `#include`ed this header that are no longer necessary.

### CODE REVIEW

File: include/deal.II/base/table.h [hunk 1]
```diff
@@ -828,6 +828,264 @@ class Table<1, T> : public TableBase<1, T>
...
+   * Accessor class template. This will be partially specialized for both
+   * values of <code>Constness</code>.
+   */
+  template <typename TableType, bool Constness, Storage storage_order>
+  class Accessor;
+
+  /**
+   * Accessor base class. Stores a const pointer when <code>Constness ==
+   * true</code> and a non-const pointer when <code>Constness == false</code>,
+   * and assumes that the data is a 2D arrays in the specified storage order.
```
Comment (bangerth) [inline include/deal.II/base/table.h]: arrays -> array

File: include/deal.II/base/table.h [hunk 2]
```diff
@@ -828,6 +828,264 @@ class Table<1, T> : public TableBase<1, T>
...
+  class AccessorBase
+  {
+  public:
+    /**
+     * Type of the stored pointer to the TransposeTable.
+     */
+    using container_pointer_type = typename std::
+      conditional<Constness, const TableType *, TableType *>::type;
+
+    using value_type = typename TableType::value_type;
```
Comment (bangerth) [inline include/deal.II/base/table.h]: can you add a comment for completeness?

File: include/deal.II/base/table.h [hunk 3]
```diff
@@ -2661,60 +2893,76 @@ namespace TransposeTableIterators
...
   }
 
 
 
-  template <typename T, bool Constness>
-  inline typename AccessorBase<T, Constness>::size_type
-  AccessorBase<T, Constness>::AccessorBase::row() const
+  template <typename TableType, bool Constness, Storage storage_order>
+  inline typename AccessorBase<TableType, Constness, storage_order>::size_type
+  AccessorBase<TableType, Constness, storage_order>::AccessorBase::row() const
```
Comment (bangerth) [inline include/deal.II/base/table.h]: I know this is pre-existing, but is there an `AccessorBase::` too many here?

File: include/deal.II/base/table.h [hunk 4]
```diff
@@ -2661,60 +2893,76 @@ namespace TransposeTableIterators
...
   }
 
 
 
-  template <typename T, bool Constness>
-  inline typename AccessorBase<T, Constness>::size_type
-  AccessorBase<T, Constness>::AccessorBase::column() const
+  template <typename TableType, bool Constness, Storage storage_order>
+  inline typename AccessorBase<TableType, Constness, storage_order>::size_type
+  AccessorBase<TableType, Constness, storage_order>::AccessorBase::column()
```
Comment (bangerth) [inline include/deal.II/base/table.h]: and here

File: include/deal.II/base/table.h [hunk 5]
```diff
@@ -2661,60 +2893,76 @@ namespace TransposeTableIterators
...
+                                                           container));
   }
 
 
 
-  template <typename T, bool Constness>
+  template <typename TableType, bool Constness, Storage storage_order>
   inline void
-  AccessorBase<T, Constness>::AccessorBase::assert_valid_linear_index() const
+  AccessorBase<TableType, Constness, storage_order>::assert_valid_linear_index()
```
Comment (bangerth) [inline include/deal.II/base/table.h]: whereas you successfully fixed it here

**Comment by drwells** [comment]
I fixed all of the suggested items and also added more documentation to each of the new classes.

File: include/deal.II/base/table.h (hunk 3 above)
Comment (drwells) [reply include/deal.II/base/table.h]: I am a bit frightened that this ever worked.

Comment (bangerth) [reply include/deal.II/base/table.h]: The name of the class is available inside the class. I'm not terribly surprised that it works, it just looks funny :-)

**Pull request merged**